Vanilla Behavior Overhaul

garrettjoecox opened the pull request:
## Vanilla Behavior Overhaul
I know this PR is large, and a bit of a mess, hoping to get some more eyes and hands on it. Things are going to be broken for a good bit, but this is the main thing I wanted to do for v3.

### What is this?
This is an overhaul of how we make changes to the game's source code. The outcomes of these changes include:
- Cutscene/time savers no longer arbitrarily tied to randomizer
- Reduce our footprint on the source code, making it easier to pull in upstream documentation, along with clear boundaries of what we've changed
- Randomizer check giving system migrated to be 2 step (asynchronous), more explanation below
- Because of the very obvious boundary between how we make changes in the source code, any mods can take advantage of any changes other mods make (for instance, someone could implement a different randomizer backend, and would require very little changes to the source code, or a mod could replace the item given for a single "check" in the vanilla game without opting into the entire randomizer system)

### How does it work?
The foundation for this PR is `GameInteractor_Should`, a new function that returns a boolean that is the result of whatever mods have opted into the hook and determined if something _should_ or _should not_ happen. It takes the following arguments:
- `GIVanillaBehavior id` - An enum that represents what behavior we are overriding. Should read like a sentence with the hook name (ignoring enum prefix), for instance "Should be eligible for light arrows" is `GI_VB_BE_ELIGIBLE_FOR_LIGHT_ARROWS`
- `bool should` - A boolean that represents how the vanilla game would behave. This is the default value, and is the value that should be returned if no mods have opted into the hook. In cases where you want to _override_ a vanilla condition, you would simply wrap the original condition in this hook (A). In cases where you want to are just wrapping code you don't want to happen that normally happens in vanilla, you would just pass in `true` here (B).
- `void* opt` - A pointer to an arbitrary value relevant to the behavior that may be useful for the consumer of the hook. In most cases this would be NULL or a pointer to the actor associated with the behavior. What this is will be documented along side it's GIVanillaBehavior enum (This is maybe a bit dangerous, we can aim for more safety here later?)

(A) Example of overriding a vanilla condition, we want to override this in rando to instead check if the player has used the deku tree blue warp:
```diff
-if (CHECK_QUEST_ITEM(QUEST_KOKIRI_EMERALD)) {
+if (GameInteractor_Should(GI_VB_MIDO_CONSIDER_DEKU_TREE_DEAD, CHECK_QUEST_ITEM(QUEST_KOKIRI_EMERALD), this)) {
    return 0x1045;
}
```
(B) Example of overriding a vanilla behavior, we don't want this cutscene to start when disabling cutscenes:
```diff
Flags_SetEventChkInf(EVENTCHKINF_LEARNED_SARIAS_SONG);
+if (GameInteractor_Should(GI_VB_PLAY_SARIAS_SONG_CS, true, this)) {
    play->csCtx.segment = SEGMENTED_TO_VIRTUAL(spot05_scene_Cs_005730);
    gSaveContext.cutsceneTrigger = 1;
    this->actionFunc = func_80AF68E4;
+}
```
On the hook consumer side, we have the `OnVanillaBehavior` hook, called with the same arguments as `GameInteractor_Should`, except the `bool should` argument is instead a pointer `bool* should` to whether or not the behavior should happen. Each consumer of this hook can change the value of `should` to whatever they want, and the final value will be returned to the caller of `GameInteractor_Should`. This hook is called in the order that the hooks are registered, so the last mod to opt into the hook will have the final say on the value of `should`, should they choose to change it. This is an example consuming the example behavior above:
```cpp
void OnVanillaBehaviorHandler(GIVanillaBehavior id, bool* should, void* opt) {
    switch (id) {
        case GI_VB_MIDO_CONSIDER_DEKU_TREE_DEAD: {
            if (IS_RANDO) {
                // In rando, we care about using the blue warp instead of obtaining the emerald
                *should = Flags_GetEventChkInf(EVENTCHKINF_OBTAINED_KOKIRI_EMERALD_DEKU_TREE_DEAD);
                break;
            }
        }
        case GI_VB_PLAY_SARIAS_SONG_CS: {
            if (CVarGetInteger("gTimeSavers.SkipCutscene.LearnSong", 0)) {
                EnSa* enSa = (EnSa*)opt;
                // func_80AF6B20 is the function Saria ends up in after the cutscene
                enSa->actionFunc = func_80AF6B20;
                *should = false;
            }
            break;
        }
    }
}
```

### 2 Step Randomizer Check System
Utilizing the system detailed above, along with other hooks, we are migrating all randomizer checks to an asynchronous 2 step system. 

The first step, usually happening within the original source code, is to set a flag and then ensure Link is done interacting with whatever actor/cutscene is associated with the check. This means every randomizer check will be closely associated with a flag in the game, preferably vanilla flags, but in some cases where flags don't exist we will create a new soh_inf (currently randomizer_inf) flag.

The second step, is a hook handler watching for flag sets, and when a flag is set that is associated with a randomizer check it queues up an item to be given.

### TODO:
- Misc Todo Items:
  - [ ] Adjustments to logic to account for RC_GIFT_FROM_SAGES/RC_LINKS_POCKET being different checks
  - [ ] Fix GI messages in cutscenes for rando (the model renders the randomized item, but the message still describes the vanilla item)
  - [ ] [annoyance] Shiek in your face first time going adult
  - [ ] Probably should happen in seperate PR, but rename randomizer_inf to soh_inf
  - [ ] Probably should happen in separate PR, Remove reduntant flag/identification info in location_list. We don't need ActorId, SceneNum, ActorParams, Flags, AND SpoilerCollectionCheckGroup.
  - [ ] Item tracker needs to be re-worked, moving different states it can be in to ItemLocation in Randomizer, falling back to flags if not in randomizer?
  - [ ] Dampe always win checkbox is disabled for rando, should just default to on for rando
  - [x] Track down all issues introduced with skipping one point cutscenes
    - [x] Forest temple rotating hallway borked/weird 
  - [ ] Skip item get animations only works on rando currently
  - [ ] What do we want the behavior to be when picking up an item on a ladder or when holding a cucco
  - [x] Sometimes in rando, the redead grave in the graveyard (where you have to play the sun song) doesn't spawn a chest
- [ ] Cutscene Skips
  - [x] All option
  - [x] Individual options
  - [x] One Point Skips 
  - [x] Skipped cutscenes should still set the time
  - [x] Ensure compatibility with entrance randomizer
  - [x] Entrance cutscenes
  - [x] Glitch-aiding cutscenes (checkbox is there, doesn't do anything yet)
  - [x] Title card toggle
  - [x] Intro cutscene 
  - [ ] Boss introductions
    - [x] Ghoma
    - [x] King Dodongo
    - [x] Barinade
    - [x] Phantom Ganon
    - [x] Volvagia
    - [x] Morpha
    - [x] Nabooru
    - [x] Twinrova
    - [x] Bongo Bongo
    - [ ] Ganondorf
    - [ ] Ganon
  - [ ] Quick Boss Deaths
    - [x] Phantom Ganon
    - [x] Nabooru
    - [ ] Twinrova
    - [ ] Ganondorf
  - [x] Learn Song
    - [x] Zelda's Lullaby
    - [x] Epona's Song
    - [x] Saria's Song
    - [x] Sun's Song
    - [x] Song of Time
    - [x] Song of Storms
    - [x] Minuet
    - [x] Bolero
    - [x] Serenade
    - [x] Requiem
    - [x] Nocturne
    - [x] Prelude
  - [x] Story cutscenes
    - [x] Lost woods bridge
    - [x] Zelda in courtyard
    - [x] Zelda Escape
    - [x] First time pulling sword
    - [x] Nabooru captured (On open silver gauntlets chest)
    - [x] Drain Well
    - [x] LACS
    - [x] Ganon Castle Collapse
    - [x] Deku Tree Blue Warp
    - [x] Dodongo's Cavern Blue Warp
    - [x] Jabu Jabu's Belly Blue Warp
    - [x] Forest Temple Blue Warp
    - [x] Fire Temple Blue Warp
    - [x] Water Temple Blue Warp
    - [x] Spirit Temple Blue Warp
    - [x] Shadow Temple Blue Warp
  - [x] Fairy Fountains
  - [x] Owl
  - [x] Misc Speedups
    - [x] Seed pouch upgrade in lost woods (this was part of https://github.com/garrettjoecox/OOT/pull/95 right?)
    - [x] Time Travel
    - [x] Opening Jabu Jabu
    - [x] Deku Scrub Theater
    - [x] Forest Temple Purple Poe Fight
    - [x] Goron crying in Gc as adult
    - [x] Wonder talk
    - [x] Door of time cutscene
    - [x] Darunia's Joy
    - [x] Rainbow bridge
    - [x] Opening royal tomb in graveyard
    - [x] Ganon trials
    - [x] Child Stealth
    - [ ] Ruto skips within jabu
- [ ] Rando 2 Step Check Migration (~700 Checks Total)
  - [x] Chests (301 Checks)
    - [x] Verify all chests are correctly flagged
  - [x] Free Standing Items (~33 Checks)
  - [x] Gold Skulltulas (144 Checks)
  - [ ] Shop Items (64 Checks)
  - [x] Business Scrub (46 Checks)
  - [x] Cows (10 Checks)
  - [x] Blue Warp (8 Checks)
  - [x] Cutscene Items
    - [x] RC_TOT_MASTER_SWORD
    - [x] RC_TOT_LIGHT_ARROWS_CUTSCENE
    - [x] RC_LW_GIFT_FROM_SARIA
    - [x] RC_HF_OCARINA_OF_TIME_ITEM
  - [x] Song Checks
    - [x] RC_SONG_FROM_IMPA
    - [x] RC_SONG_FROM_MALON
    - [x] RC_SONG_FROM_SARIA
    - [x] RC_SONG_FROM_ROYAL_FAMILYS_TOMB
    - [x] RC_SONG_FROM_OCARINA_OF_TIME
    - [x] RC_SONG_FROM_WINDMILL
    - [x] RC_SHEIK_IN_FOREST
    - [x] RC_SHEIK_IN_CRATER
    - [x] RC_SHEIK_IN_ICE_CAVERN
    - [x] RC_SHEIK_AT_COLOSSUS
    - [x] RC_SHEIK_IN_KAKARIKO
    - [x] RC_SHEIK_AT_TEMPLE
  - [x] Trade Sequence Actor Checks
    - [x] RC_LW_TRADE_COJIRO
    - [x] RC_LW_TRADE_ODD_POTION
    - [x] RC_LH_TRADE_FROG
    - [x] RC_KAK_TRADE_ODD_MUSHROOM
    - [x] RC_KAK_TRADE_POCKET_CUCCO
    - [x] RC_GV_TRADE_SAW
    - [x] RC_DMT_TRADE_BROKEN_SWORD
    - [x] RC_ZD_TRADE_PRESCRIPTION
    - [x] RC_DMT_TRADE_EYEDROPS
    - [x] RC_DMT_TRADE_CLAIM_CHECK
  - [x] Merchant Actor Checks
    - [x] RC_WASTELAND_BOMBCHU_SALESMAN
    - [x] RC_GC_MEDIGORON
    - [x] RC_KAK_GRANNYS_SHOP
    - [x] RC_ZR_MAGIC_BEAN_SALESMAN
  - [x] Frog Songs (7 Checks)
  - [ ] Fishsanity (ottosparks maybe)
  - [x] Beehives (bria)
  - [ ] Actor Checks
    - [x] RC_LW_SKULL_KID
    - [x] RC_LW_OCARINA_MEMORY_GAME (Taw)
    - [x] RC_LW_TARGET_IN_WOODS
    - [x] RC_DEKU_THEATER_SKULL_MASK
    - [x] RC_DEKU_THEATER_MASK_OF_TRUTH
    - [ ] RC_LH_CHILD_FISHING (ottosparks maybe)
    - [ ] RC_LH_ADULT_FISHING (ottosparks maybe)
    - [x] RC_LH_LAB_DIVE
    - [x] RC_LH_UNDERWATER_ITEM (bria)
    - [x] RC_LH_SUN (bria)
    - [x] RC_GF_HBA_1000_POINTS (bria)
    - [x] RC_GF_HBA_1500_POINTS (bria)
    - [x] RC_GF_GERUDO_MEMBERSHIP_CARD
    - [x] RC_GF_NORTH_F1_CARPENTER (Freestanding?) (bria tested)
    - [x] RC_MARKET_TREASURE_CHEST_GAME_REWARD (bria)
    - [x] RC_MARKET_BOMBCHU_BOWLING_FIRST_PRIZE (bria)
    - [x] RC_MARKET_BOMBCHU_BOWLING_SECOND_PRIZE (bria)
    - [x] RC_MARKET_LOST_DOG (taw)
    - [x] RC_MARKET_SHOOTING_GALLERY_REWARD
    - [x] RC_HC_MALON_EGG
    - [x] RC_HC_ZELDAS_LETTER
    - [x] RC_KAK_10_GOLD_SKULLTULA_REWARD (bria)
    - [x] RC_KAK_MAN_ON_ROOF
    - [x] RC_KAK_SHOOTING_GALLERY_REWARD
    - [x] RC_KAK_ANJU_AS_ADULT
    - [x] RC_KAK_ANJU_AS_CHILD
    - [x] RC_GRAVEYARD_DAMPE_GRAVEDIGGING_TOUR
    - [x] RC_GC_ROLLING_GORON_AS_CHILD
    - [x] RC_GC_ROLLING_GORON_AS_ADULT
    - [x] RC_GC_DARUNIAS_JOY
    - [x] RC_ZD_DIVING_MINIGAME (bria)
    - [x] RC_ZD_KING_ZORA_THAWED
    - [x] RC_LLR_TALONS_CHICKENS (bria)

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1445531243.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1445534681.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1445538786.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1445539169.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1445542666.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1445571463.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1445758870.zip)
<!--- section:artifacts:end -->

fossifousacid commented on soh/soh/Enhancements/randomizer/3drando/fill.cpp:
```diff
@@ -1123,7 +1123,19 @@ int Fill() {
-    //Add prices for scrubsanity, this is unique to SoH because we write/read scrub prices to/from the spoilerfile.
+    //Add default prices to scrubs
+    for (size_t i = 0; i < Rando::StaticData::scrubLocations.size(); i++) {
+      if (Rando::StaticData::scrubLocations[i] == RC_LW_DEKU_SCRUB_NEAR_BRIDGE || Rando::StaticData::scrubLocations[i] == RC_LW_DEKU_SCRUB_GROTTO_FRONT) {
```
Mind if I take a crack at refactoring this for legibility? There's probably little benefit to using C style syntax here.

garrettjoecox replied:
Have at it!

garrettjoecox committed:
Vanilla Behavior Overhaul

Co-authored-by: jordanpg <jordanpg@users.noreply.github.com>

Tawling committed:
GameInteractor hooks for adult and child shooting gallery

Tawling committed:
Man On Roof

garrettjoecox committed:
Merge pull request #83 from Tawling/vbo-shooting-gallery

GameInteractor hooks for adult and child shooting gallery

Tawling committed:
fix include

briaguya0 committed:
fix the torch

garrettjoecox committed:
Merge pull request #90 from briaguya-ai/fix-the-torch

fix the torch

garrettjoecox committed:
Merge pull request #85 from Tawling/vbo-man-on-roof

GI VB for Man On Roof

Tawling committed:
Skull Kid (#86)

Co-authored-by: Garrett Cox <garrettjcox@gmail.com>

garrettjoecox committed:
More boss intro skips

garrettjoecox committed:
Re-add some UI I accidentally removed

garrettjoecox committed:
Migrate song of storms check

garrettjoecox committed:
Migrate okarina_tag changes

garrettjoecox committed:
Merge branch 'develop-rando' of garrettjoecox.github.com:HarbourMasters/Shipwright into develop-rando-changes

garrettjoecox committed:
Enable timesaver enhancements by default on rando

garrettjoecox committed:
Various adjustments to goron related vanilla behavior hooks

garrettjoecox committed:
More 1PC skips

garrettjoecox committed:
Migrate great fairy checks

briaguya0 committed:
pull in docs (#97)

* rename `func_8002DF54` to `Player_SetCsActionWithHaltedActors`
* rename `func_8002F434` to `Actor_OfferGetItem`

briaguya0 committed:
vb target in woods (#95)

* do some target in woods VB stuff

* skip deku scrub hopping out animation

* replace a couple IS_RANDOs

* cleanup

briaguya0 committed:
vb deku theater (#96)

garrettjoecox committed:
Merge branch 'develop-rando' of garrettjoecox.github.com:HarbourMasters/Shipwright into develop-rando-changes

briaguya0 committed:
pull docs: talon function names (#98)

briaguya0 committed:
vb diving game (#100)

briaguya0 committed:
vb talon's chickens (#99)

briaguya0 committed:
vb skull rewards (#101)

briaguya0 committed:
vb horseback archery (#102)

* vb horseback archery

* Update soh/soh/Enhancements/randomizer/hook_handlers.cpp

---------

Co-authored-by: Garrett Cox <garrettjcox@gmail.com>

garrettjoecox committed:
Merge branch 'develop-rando' of garrettjoecox.github.com:HarbourMasters/Shipwright into develop-rando-changes

briaguya0 committed:
vb underwater item (#104)

* underwater working pre func_80B858B4 replacement

* underwater item update handled by hookhandlers

* remove commented out code

* remove unused should

briaguya0 committed:
vb lh sun (#106)

* underwater working pre func_80B858B4 replacement

* underwater item update handled by hookhandlers

* vb lh sun

* remove unused should

briaguya0 committed:
vb treasure chest game (#103)

briaguya0 committed:
vb ocarina memory (#107)

* Ocarina Memory Game

* vb memory game

---------

Co-authored-by: Taw <Tawling@users.noreply.github.com>

briaguya0 committed:
vb richard (#108)

* Lost Dog + Custom Hook Handler

* vb richard

* fix the merge

---------

Co-authored-by: Taw <Tawling@users.noreply.github.com>

briaguya0 committed:
vb beehives (#109)

* vb beehives

* format

briaguya0 committed:
vb bombchu bowling (#110)

* vb bombchu bowling

* don't need it

* it's drawing the right thing but it's tiny

* scale

* no need to should, don't rescale every draw

* slightly safer

briaguya0 committed:
vb gerudo (#111)

* vb gerudo

* clean up

garrettjoecox committed:
Migrate ganon trials

garrettjoecox committed:
Migrate rainbow bridge

garrettjoecox committed:
Migrate purple poe faster fight

garrettjoecox committed:
Finish glitch aiding cutscenes toggle

garrettjoecox committed:
Merge branch 'develop-rando' of garrettjoecox.github.com:HarbourMasters/Shipwright into develop-rando-changes

briaguya0 committed:
fix: add location access for treasure chest game items

vb treasure chest game added new RCs for the treasure chest game (now there are RCs for each of the 5 keys and each of the 5 non key chests), but the location area table only had the keys, this adds the items

briaguya0 committed:
fix: gift from sages location access

garrettjoecox committed:
Merge pull request #113 from briaguya-ai/treasure-chest-game-location-access

fix: add location access for treasure chest game items and gift from sages

briaguya0 committed:
vb tower escape

garrettjoecox committed:
Merge pull request #114 from briaguya-ai/vb-tower-escape-no-new-should

vb tower escape

garrettjoecox committed:
More time cutscene skip fixes

garrettjoecox committed:
Migrate child stealth

garrettjoecox committed:
Fix shadow & spirit rewards, remove remaning usage of duplicate dungeon done flags

briaguya0 committed:
move `RawJson` to LUS (#4063)

garrettjoecox committed:
Fix for some chests not appearing

garrettjoecox committed:
Merge branch 'develop-rando' of garrettjoecox.github.com:HarbourMasters/Shipwright into develop-rando-changes

garrettjoecox committed:
Point LUS to fork temporarily for dl fix

garrettjoecox committed:
Fix forest temple entrance cs skip

garrettjoecox committed:
Merge branch 'develop-rando' of garrettjoecox.github.com:HarbourMasters/Shipwright into develop-rando-changes

garrettjoecox committed:
Add missing RHT entries

garrettjoecox committed:
Merge branch 'develop-rando' of garrettjoecox.github.com:HarbourMasters/Shipwright into develop-rando-changes

briaguya0 committed:
latest lus main

garrettjoecox committed:
Merge pull request #117 from briaguya-ai/latest-lus-main

latest lus main

briaguya0 committed:
don't skip item give animation in chu bowling

garrettjoecox committed:
Merge pull request #118 from briaguya-ai/chu-bowling-dont-skip

don't skip item give animation in chu bowling

leggettc18 committed:
Adds Quick Boss Death checkbox and implements Phantom Ganon's quick death (#119)

* Adds Quick Boss Death checkbox and implements Phantom Ganon's quick death.

* Clarifies relocation comment.

briaguya0 committed:
vb deku stick cheat (#115)

* vb deku stick cheat

* it's the 21st now

---------

Co-authored-by: Garrett Cox <garrettjcox@gmail.com>

Archez committed:
explicitly include unordered_map header (#122)

leggettc18 committed:
Conversion of Nabooru Quick Death to VB hook (#120)

* Nabooru quick death conversion

* Fix brackets, not sure exactly how this was working before.

garrettjoecox committed:
Merge branch 'develop-rando' of garrettjoecox.github.com:HarbourMasters/Shipwright into develop-rando-changes

garrettjoecox committed:
Fix two minor issues from merge
